server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java

```diff
@@ -553,13 +594,19 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                 builder.field("path_unmatch", pathUnmatch);
             }
         }
-        // We have more than one types when (1) `match_mapping_type` is "*", and (2) match and/or path_match are defined but
-        // not `match_mapping_type`. In the latter the template implicitly accepts all types and we don't need to serialize
-        // the `match_mapping_type` values.
-        if (xContentFieldTypes.length > 1 && match.isEmpty() && pathMatch.isEmpty()) {
-            builder.field("match_mapping_type", "*");
-        } else if (xContentFieldTypes.length == 1) {
-            builder.field("match_mapping_type", xContentFieldTypes[0]);
+        if (matchMappingType.isEmpty() == false) {
+            if (matchMappingType.size() == 1) {
+                builder.field("match_mapping_type", matchMappingType.get(0));
+            } else {
+                builder.field("match_mapping_type", matchMappingType);
+            }
+        }
+        if (unmatchMappingType.isEmpty() == false) {
+            if (unmatchMappingType.size() == 1) {
+                builder.field("unmatch_mapping_type", unmatchMappingType.get(0));
+            } else {
+                builder.field("unmatch_mapping_type", unmatchMappingType);
+            }
```

Comment:
Logic seems similar...maybe extract a method just passing a list and a string?

Reply:
let me add that